ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -53,39 +49,58 @@ final class InMemoryLedgerReaderWriter(
     extends LedgerWriter
     with LedgerReader {
 
-  private val engine = Engine()
-
   private val currentState = new InMemoryState()
 
-  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
-    Future {
-      val submission = Envelope
-        .openSubmission(envelope)
-        .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
-      currentState.synchronized {
-        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
-          submission.getInputDamlStateList.asScala
-            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
-        val entryId = sequentialLogEntryId.next()
-        val (logEntry, damlStateUpdates) =
-          KeyValueCommitting.processSubmission(
-            engine,
-            entryId,
-            currentRecordTime(),
-            LedgerReader.DefaultConfiguration,
-            submission,
-            participantId,
-            stateInputs,
-          )
-        val stateUpdates = damlStateUpdates.map {
-          case (damlStateKey, value) => damlStateKey.toByteString -> value
+  private class InMemoryLedgerStateAccess(theParticipantId: ParticipantId)
+      extends LedgerStateAccess {
+    override def inTransaction[T](body: LedgerStateOperations => Future[T]): Future[T] =
+      Future {
+        currentState.state.synchronized {
+          body(new InMemoryLedgerStateOperations)
         }
```

Comment:
What does it mean to synchronize over a `Future`? I think you need to use an explicit lock and unlock on a mutex here, otherwise the lock will be released when the future is returned, not when it completes.

Something like:

```scala
Future(lock() /* example; not a thing */)
  .flatMap(lock =>
    body(new InMemoryLedgerStateOperations)
      .transform { result =>
         lock.unlock()
         result
      })
```

Don't use a `ReentrantLock` because that forces you to lock and unlock on the same thread. You'll probably need to use a `Semaphore`.

Reply:
It doesn't make any sense -- changed according to your suggestion. I.e., state was only locked while the parent `Future` was running hence it did not have the effect needed.